onnxruntime/core/optimizer/selectors_actions/selector_action_transformer.cc

```diff
@@ -256,7 +256,14 @@ static Status SetOpSinceVersionForProducedNodes(NodeIndex pre_action_max_num_nod
 
 Status SelectorActionTransformer::ApplySavedRuntimeOptimizations(
     Graph& graph, bool& modified, int graph_level, const logging::Logger& logger) const {
-  for (auto& node : graph.Nodes()) {
+  GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
```

Comment:
Same question there. If we don't need to pay the cost of the GraphViewer construction we shouldn't.

Reply:
I thought since in the loop, there are logic removing the node, it might be tricky to use a container iterator here.

Do you think it is a problem, if not, I can revert the change back.

Reply:
It never broke before so I don't think it's a problem. :) 

The node pointers in the Graph that are being iterated are in a vector. Removing the node sets the slot to a nullptr. If we actually removed the entry all the node index values would be invalidated as well as they're indexes into that vector.

Reply:
aha, just to make sure I understand it correctly, "Removing the node sets the slot to a nullptr". Will we still get a nullptr Node once a "future" node is removed in current iteration. I guess not, right, otherwise, it's a problem for the existing code.

Reply:
Never mind, I got the answer. 

  // Wrapper of Graph nodes to provide iteration services that hide nullptr entries
  GraphNodes iterable_nodes_{nodes_};

Will revert the changes. :)

Reply:
That iterator will filter out the nullptr entries. 

If you iterate the topologically sorted nodes via GraphViewer it will not (at least not currently) filter them out, as it's iterating NodeIndex values in a vector (GraphViewer::nodes_in_topological_order_).  

GraphViewer::nodes_in_topological_order_ is populated when the GraphViewer instance is created and never updated, as it's a GraphViewer that can't modify the Graph. That means it will be out of sync when the Graph is directly modified. 

It doesn't make a lot of sense to me to be using GraphViewer in optimizers that modify the graph, especially given the Graph class has a nodes_in_topological_order_ member.

Reply:
I see, thanks a lot!